src/vs/workbench/api/browser/mainThreadLanguageFeatures.ts

```diff
@@ -330,8 +331,20 @@ export class MainThreadLanguageFeatures extends Disposable implements MainThread
 	// --- quick fix
 
 	$registerQuickFixSupport(handle: number, selector: IDocumentFilterDto[], metadata: ICodeActionProviderMetadataDto, displayName: string, supportsResolve: boolean): void {
+
 		const provider: languages.CodeActionProvider = {
 			provideCodeActions: async (model: ITextModel, rangeOrSelection: EditorRange | Selection, context: languages.CodeActionContext, token: CancellationToken): Promise<languages.CodeActionList | undefined> => {
+
+				// if provideCodeActions was trigged by an INVOKE, do not filter notebook CA providers
+				if (context.trigger !== languages.CodeActionTriggerType.Invoke) {
```

Comment:
> if provideCodeActions was trigged by an INVOKE, do not filter notebook CA providers

Can you explain what that is? I find this very confusing and IMO it makes it confusing to register a notebook only code action provider. If I want to target notebook cells I can still narrow down on them with the scheme, notebook-type and language selector parts

Reply:
I think this is somewhat unique to notebook CodeActions, as we only ever run them on save. Without this (presuming you keep the below check for notebook kind providers), you end up filtering out notebook CodeActions at all times -- including on save. 

If we are restricting Notebook CodeActions to only run on save, and don't want the providers called here, we have to distinguish between their `TriggerType` was my thinking. The `context.trigger` will be `automatic` in most cases -- lightbulb, document loading, etc -- but is `invoke` when run on save or `ctrl+.`

So:
- automatic triggers (lightbulb, document load) should NOT call against `notebook.` providers
- invoked triggers (on save, `ctrl+.`) should call against them according to their selectors

This was my thinking during the changes, feel free to correct me if there's any spots I'm off track. Also just tweaked the first condition for a bit more logical readability.

Reply:
> invoked triggers (on save, ctrl+.) should call against them according to their selectors

I don't think that's a good idea because it makes implementing a notebook-wide code action provider very cumbersome. When making calls from the renderer we need to know if they happen for a notebook (instead of a notebook cell) like saving a notebook or some future notebook-wide code action gesture.